added jwt auth

iclegane opened the pull request.

iclegane committed:
added jwt auth

iclegane committed:
fixed README.md

iclegane committed:
fixed validateMiddleware.ts

iclegane commented on server/src/routes/Auth.ts:
```diff
@@ -0,0 +1,118 @@
+
+const router = Router();
+
+router.post('/login', validate(loginSchema), async (req, res) => {
```
Думаю еще сохранять refreshToken в бд, чтобы можно было со всех устройств логаут сделать

iclegane committed:
added front auth

iclegane committed:
fixed types

iclegane committed:
- fixed port

iclegane committed:
fixed data func

iclegane commented on client/src/pages/CreateAccount/components/CreateAccountForm/CreateAccountForm.tsx:
```diff
@@ -1,64 +1,106 @@
+    formState: { errors },
+  } = useForm<FormValues>()
+
+  const onSubmit: SubmitHandler<FormValues> = async (data) => {
```
1

iclegane committed:
- added logout to board menu

iclegane committed:
added card api

iclegane committed:
fixed package.json

iclegane committed:
added node v

iclegane committed:
added ws

iclegane committed:
added WebSocketServer

Ayub-Begimkulov commented on client/src/api/axios.ts:
```diff
@@ -0,0 +1,40 @@
+
+        error.config.headers.Authorization = `Bearer ${newAccessToken}`
+        return api.request(error.config)
+      } catch (_e) {
```
можно по идее просто `catch {}` писать

Ayub-Begimkulov commented on client/src/api/axios.ts:
```diff
@@ -0,0 +1,40 @@
+  async (error) => {
+    if (error.response?.status === 401) {
+      try {
+        const refreshResponse = await axios.post(
+          API_PATH.REFRESH,
+          {},
+          { withCredentials: true }
+        )
+        const newAccessToken = refreshResponse.data.accessToken
+        localStorage.setItem(ACCESS_TOKEN_KEY, newAccessToken)
```
а рефреш в куках, да?

iclegane replied:
да

Ayub-Begimkulov commented on client/src/constants/index.ts:
```diff
@@ -0,0 +1,10 @@
+  CREATE = 'auth/create',
+  LOGOUT = 'auth/logout',
+  REFRESH = 'auth/refresh',
+}
```
вот тут вообще плохое использование enum'ов. то есть даже если споры про нужны или не нужные енумы оставить - то это не тот кейс, где их пременять надо. по сути они хорошо ложатся на однотипные данные с разными вариантами.

status - loading/success/error
gender - male/female

iclegane replied:
Я исходил с того, что адрес или ручка может поменяться, ну и достаточно значение только поменять. 

Я так понимаю ts в любом случае enum в константу с объектом превратит и будет что то типо:

```
Const api = {
…
} as const
```

Не пойму тогда кейсы с enum, где бы не подошел обычны объект)

Ayub-Begimkulov commented on client/src/context/AuthContext.tsx:
```diff
@@ -0,0 +1,76 @@
+  }, [])
+
+  const value = useMemo(
+    () => ({ accessToken, login, logout, isLogged }),
```
а зачем в контексте хранить accessToken? он же сразу у нас подставляется из интерцептора)
+ можно всегда из localStorage достать, если надо...

Ayub-Begimkulov commented on client/src/context/AuthContext.tsx:
```diff
@@ -0,0 +1,76 @@
+    localStorage.getItem(ACCESS_TOKEN_KEY)
+  )
+
+  const isLogged = Boolean(accessToken)
```
вот тут тоже странная немного логика. по идее это надо определять же на основе запроса при первом заходе. может я сам дичь поставил в accessToken.

Ayub-Begimkulov commented on client/src/pages/CreateAccount/components/CreateAccountForm/CreateAccountForm.tsx:
```diff
@@ -1,64 +1,101 @@
-import React from 'react'
+import React, { useState } from 'react'
+import { Controller, type SubmitHandler, useForm } from 'react-hook-form'
```
все-таки решил заюзать? хахах

iclegane replied:
Да, так побыстрее) А то на другие кейсы отвлекаюсь

Ayub-Begimkulov commented on client/src/pages/CreateAccount/components/CreateAccountForm/CreateAccountForm.tsx:
```diff
@@ -1,64 +1,101 @@
-    e.preventDefault()
+  const onSubmit: SubmitHandler<FormValues> = async (data) => {
+    setIsLoading(true)
+    api.post(API_PATH.CREATE, { ...data }).then(() => {
```
а зачем тут копирование `{...data}`?

Ayub-Begimkulov commented on client/src/pages/CreateAccount/components/CreateAccountForm/CreateAccountForm.tsx:
```diff
@@ -1,64 +1,101 @@
+    api.post(API_PATH.CREATE, { ...data }).then(() => {
+      navigate('/login')
+    })
+    setIsLoading(false)
```
а вот тут await не вижу, разве сработает лоудер то?

Ayub-Begimkulov commented on client/src/pages/CreateAccount/components/CreateAccountForm/CreateAccountForm.tsx:
```diff
@@ -1,64 +1,101 @@
-        onChange={login.handleChange}
-        name={'login'}
-        placeholder='Login'
+      <Controller
+        name='login'
+        control={control}
+        rules={{ required: 'Login is required' }}
+        render={({ field }) => (
+          <Input
+            {...field}
+            placeholder='Login'
+            status={errors.login ? 'error' : 'default'}
+          />
+        )}
```
а ты разве не имитируешь пропсы нативного инпута? через register не работало?

iclegane replied:
По формам тогда в некст мре внесу правки

Ayub-Begimkulov commented on client/src/pages/Login/components/LoginForm/LoginForm.tsx:
```diff
@@ -1,51 +1,91 @@
+      navigate('/board')
+    })
+    setIsLoading(false)
+  }
```
тут тоже чот авайт не видно)

Ayub-Begimkulov commented on server/nodemon.json:
```diff
@@ -0,0 +1,6 @@
+  "watch": ["src"],
+  "ignore": ["dist", "node_modules"],
+  "ext": "ts",
+  "exec": "tsx src/index.ts"
```
а почему просто tsx watch не использовать? зачем nodemon?)

Ayub-Begimkulov commented on server/src/dto/ResponseDTO.ts:
```diff
@@ -0,0 +1,22 @@
+// TODO: Подумать
+export class SuccessResponse<T = unknown> {
+    readonly status = 'success';
+    readonly message?: string;
```
а message нужен разве?)

Ayub-Begimkulov commented on server/src/index.ts:
```diff
@@ -1,15 +1,90 @@
+        return ws.close(1011, ERROR_MESSAGES.SERVER_ERROR);
+    }
+    const parsedUrl = new URL(url, 'http://localhost');
+    const token = parsedUrl.searchParams.get('accessToken');
```
так подожди, а почему через search передается?

Ayub-Begimkulov commented on server/src/index.ts:
```diff
@@ -1,15 +1,90 @@
+// Todo: вынести
+mongoose.connect(MONGO_URI)
+    .then(() => console.log('MongoDB connected'))
+    .catch(err => console.log(err));
```
в идеале бы да, для базы сделать какой-то класс обертку)

Ayub-Begimkulov commented on server/src/middleware/validateMiddleware.ts:
```diff
@@ -0,0 +1,19 @@
+
+export const validate = (schema: AnySchema): RequestHandler => async (req, res, next) => {
+    try {
+        req.body = await schema.validate(req.body, { abortEarly: false, stripUnknown: true });
```
а зачем тут мутировать?

Ayub-Begimkulov commented on server/src/models/Cards.ts:
```diff
@@ -0,0 +1,16 @@
+
+const cardSchema = new mongoose.Schema(
+    {
+        id: { type: String, default: uuidv4, unique: true },
```
а для чего самому тут id добавлять? если в модель и так по дефолту добавляет монгуз? тут еще uuid может дать занчение ж, которое уже есть в базе

Ayub-Begimkulov reviewed (commented):
я везде не писал, но надо бы на беке обработку ошибок проверить, что она там норм все кейсы на клиент отдеает.

iclegane committed:
- fixed auth

iclegane committed:
- fixed auth

iclegane committed:
- fixed auth

iclegane committed:
- fixed auth

iclegane committed:
added ws

iclegane committed:
added config

iclegane committed:
removed nodemon

iclegane committed:
updated cardSchema

iclegane committed:
- added ws

iclegane committed:
added prettier&lint

iclegane committed:
lint fix

iclegane committed:
- added todo

iclegane committed:
fixed card id in model

iclegane committed:
Refactoring

Ayub-Begimkulov commented on client/src/api/axios.ts:
```diff
@@ -1,38 +1,41 @@
+      } catch {
+        return Promise.reject(error)
+      } finally {
+        isRetried = true
```
у тебя с такой схемой только 1 раз ретрай будет работать. имею в виду на все приложение один раз) так как isRetried после первого фейла станет true и уже все)

iclegane committed:
fix ws

iclegane committed:
- added CursorSharing

iclegane committed:
- added register react hook form

Ayub-Begimkulov commented on client/src/context/AuthContext.tsx:
```diff
@@ -2,68 +2,63 @@ import React, {
+      }
+    }
+    checkAuth()
+  }, [])
```
странно такое через рефреш проверять. обычно либо данные юзера тянут (если они есть / нужны) либо отдельный делают ендпоинт для этого

Ayub-Begimkulov replied:
а на время проверки лодуер не нужен?

Ayub-Begimkulov commented on client/src/service/Auth.ts:
```diff
@@ -0,0 +1,64 @@
+
+  getToken(): string | null {
+    return localStorage.getItem(ACCESS_TOKEN_KEY)
+  }
```
а setToken сразу сюда же не поставил?)

Ayub-Begimkulov replied:
а хотя понял, у тебя через логин сеттится

Ayub-Begimkulov replied:
вообще в проде еще локал сторедж отключен может быть)) но это через отдельный сервис лучше сделать уже и в отедльной задаче

Ayub-Begimkulov commented on client/src/service/WebSocket.ts:
```diff
@@ -0,0 +1,108 @@
+    while (this.messageQueue.length && this.isReady()) {
+      this.socket?.send(this.messageQueue.shift()!)
+    }
+  }
```
n^2

Ayub-Begimkulov commented on client/src/service/WebSocket.ts:
```diff
@@ -0,0 +1,108 @@
+    this.connect()
+  }
+
+  public onMessage = (callback: Callback) => {
```
а почему часть методов через стрелки, а часть как обычные методы?

Ayub-Begimkulov commented on server/src/index.ts:
```diff
@@ -1,90 +1,33 @@
+app.use('/api/v1/board', Board)
+
+// --- Start server ---
+app.listen(CONFIG.PORT, () => {
```
у меня проект не запускается, вот такую ошибку дает (я отрубал процесс на 5000 порту, но при перезапуске снова ошибка)

```
Emitted 'error' event on Server instance at:
    at emitErrorNT (node:net:1844:8)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21) {
  code: 'EADDRINUSE',
  errno: -48,
  syscall: 'listen',
  address: '::',
  port: 5000
}
```

Ayub-Begimkulov commented on server/src/middleware/authMiddleware.ts:
```diff
@@ -1,28 +1,33 @@
-    req.userId = payload.id;
-    next();
-};
+  req.userId = payload.id
```
вообще вроде такое можно было расширять через .d.ts файл, чтобы там userId?: number был, если мне память не изменяет

Ayub-Begimkulov commented on server/src/routes/Auth.ts:
```diff
@@ -1,118 +1,129 @@
-        }
+router.post('/logout', async (_req, res) => {
+  try {
+    res.clearCookie('refreshToken', {
+      httpOnly: true,
+      secure: false,
+      sameSite: 'strict',
+    })
```
а тут разве нужны параметры?)

Ayub-Begimkulov replied:
из интересного кстати знаешь как куки чистятся в браузере?

Ayub-Begimkulov commented on server/src/routes/Auth.ts:
```diff
@@ -1,118 +1,129 @@
+  }
+})
+
+export default router
```
а почему решил через дефолтные экспорты работать? не оч тема же

Ayub-Begimkulov commented on server/src/service/Mongo.ts:
```diff
@@ -0,0 +1,24 @@
+import mongoose from 'mongoose'
+
+import logger from '../logger/index.js'
+
+export class Mongo {
+  constructor(url?: string) {
+    if (!url) {
+      throw new Error('MongoDB URI is missing')
+    }
+
+    this.connect(url)
+  }
+
+  private connect = async (url: string) => {
+    try {
+      await mongoose.connect(url)
+
+      logger.info(`🚀 [MongoDB] Connection established`)
+    } catch (error) {
+      logger.error(`🛑 [MongoDB] Connection failed:`, error)
+      throw error
+    }
+  }
```
если честно не пойму, почему это не функция? тут класс же не нужен)

Ayub-Begimkulov commented on server/src/service/WS.ts:
```diff
@@ -0,0 +1,131 @@
+
+      if (type === 'end') {
+        this.saveCardPosition(id, position)
+      }
```
очень стремная абстракция, тут же у тебя логика вебсокет сервера. а уже обработка конкретных мессаджей должна отдельно делаться...

Ayub-Begimkulov commented on server/src/service/WS.ts:
```diff
@@ -0,0 +1,131 @@
+    } catch (error) {
+      logger.warning('Error saving card:', error)
+    }
+  }
```
ну и вот этот метод да, должен вне класса быть

Ayub-Begimkulov commented on client/src/pages/Board/Board.tsx:
```diff
@@ -83,18 +84,27 @@ export const Board: React.FC = () => {
+      const name = 'Card'
+      const x = (-position.x + (window.innerWidth - cardWidth) / 2) / zoom
+      const y = (-position.y + (window.innerHeight - cardHeight) / 2) / zoom
+      const response = await api.post('/board/card', { name, x, y })
```
а почему не оптимистик?)

Ayub-Begimkulov replied:
тип из-за сокетов? чтобы сразу по ид подписывался?

Ayub-Begimkulov commented on client/src/components/Card/Card.tsx:
```diff
@@ -27,20 +28,32 @@ export const Card = memo(
     const [tempPosition, setTempPosition] = useState<Position | null>(null)
     const [dragging, setDragging] = useState(false)
 
+    const idCb = useLatest(id)
+    const tempPositionCb = useLatest(tempPosition)
```
а почему Cb? это ж не колбэки) Ref было бы норм, либо latestId/latestTempPosition

Commit:
fix package-lock.json

Commit:
fix README.md

Commit:
fix auth methods

Commit:
added auth check method

Commit:
fix n^2 complexity

Commit:
fix default routes

Commit:
fix default routes

Commit:
separated the web socket logic

Commit:
refactoring vars name

Commit:
fixed init functions

Commit:
added loader indicator

Commit:
loader indicator refactoring

Commit:
fixed:
Делать редирект с / на /board или /login в зависимости от страницы
Делать редирект на логин после создания аккаунта
Нормальная обработка ошибок для форм в UI (показывать ошибки под формой)
WS.socket on message — там может быть не тоько invalid json ошибка, но и любая другая в колбэке
Нет нормального обновление jwt через рефреш токен)

Commit:
fixed

iclegane committed:
Merge pull request #8 from iclegane/fix/JWT-auth-2

fixed:

iclegane committed:
Merge pull request #7 from iclegane/fix/JWT-auth

Fix/jwt auth
